lib/manager/cargo/artifacts.js

```diff
@@ -1,27 +1,126 @@
+const upath = require('upath');
+const process = require('process');
+const fs = require('fs-extra');
+const { exec } = require('child-process-promise');
+
 module.exports = {
   getArtifacts,
 };
 
-async function getArtifacts(cargoTomlFileName) {
-  await logger.debug(`cargo.getArtifacts(${cargoTomlFileName})`);
-  let cargoLockFileName;
-  try {
-    cargoLockFileName = cargoTomlFileName.replace(/\.toml$/, '.lock');
-    logger.debug('Updating ' + cargoLockFileName);
-    // TODO: Update cargo lock file
+// TODO: Make docker binarySource work
+async function getArtifacts(
+  packageFileName,
+  updatedDeps,
+  newPackageFileContent,
+  config
+) {
+  await logger.debug({ config }, `cargo.getArtifacts(${packageFileName})`);
+  if (updatedDeps === undefined || updatedDeps.length < 1) {
+    logger.debug('No updated cargo deps - returning null');
     return null;
-  } catch (err) {
+  }
+  if (!config.gitFs) {
     logger.warn(
-      { err, message: err.message },
-      'Failed to update Cargo lock file'
+      'Cargo lock updating requires gitFs to be enabled by the Renovate administrator.'
     );
+    return null;
+  }
+  const lockFileName = 'Cargo.lock';
+  const existingLockFileContent = await platform.getFile(lockFileName);
+  if (!existingLockFileContent) {
+    logger.debug('No Cargo.lock found');
+    return null;
+  }
+  const localPackageFileName = upath.join(config.localDir, packageFileName);
+  const localLockFileName = upath.join(config.localDir, lockFileName);
+  let stdout;
+  let stderr;
+  try {
+    await fs.outputFile(localPackageFileName, newPackageFileContent);
+    logger.debug('Updating ' + lockFileName);
+    for (let i = 0; i < updatedDeps.length; i += 1) {
+      const dep = updatedDeps[i];
+      // Update dependency `${dep}` in Cargo.lock file corresponding to Cargo.toml file located
+      // at ${localPackageFileName} path
+      const cmd = `cargo update --manifest-path ${localPackageFileName} --package ${dep}`;
+      // if (config.binarySource === 'docker') {
+      //   cmd = `docker run renovate/cargo update --manifest-path ${localPackageFileName} --package ${dep}`;
+      // }
+      const startTime = process.hrtime();
+      try {
+        ({ stdout, stderr } = await exec(`${cmd}`, {
+          shell: true,
+        }));
+      } catch (err) /* istanbul ignore next */ {
+        // Two different versions of one dependency can be present in the same
+        // crate, and when that happens an attempt to update it with --package ${dep}
+        // key results in cargo exiting with error code `101` and an error mssage:
+        // "error: There are multiple `${dep}` packages in your project".
+        //
+        // If exception `err` was caused by this, we execute `updateAll` function
+        // instead of returning an error. `updateAll` function just executes
+        // "cargo update --manifest-path ${localPackageFileName}" without the `--package` key.
```

Comment:
Does this mean that every package gets updated, not just the ones we want?

Reply:
Also, do you have an example of this type of file?

Reply:
> Does this mean that every package gets updated, not just the ones we want?

Yes, it updates all packages in `Cargo.lock`, which are out of sync with `Cargo.toml`.

> Also, do you have an example of this type of file?

Yes: https://gitlab.com/nchashch/cargo_lock
If you try running
```
cargo update -p log
```
it will error out.

Reply:
Why would it error? I only see `log` in there once

Reply:
BTW if I clone that repo and run `cargo install` then it errors, e.g.

```
error[E0603]: module `stdio` is private
   --> /Users/rhys/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.1.10/src/lib.rs:226:36
    |
226 |     handle: LineBufferedWriter<io::stdio::StdWriter>,
    |
```

Am I doing something wrong?

Reply:
`cargo update -p log` fails as you suggest. It's possible to get it to succeed if I run `cargo update -p log:0.1.0` (i.e. the old value).

`cargo update` works, but it also updates unrelated packages such as `libc`.

e.g.

```diff
 [[package]]
 name = "libc"
-version = "0.2.48"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-
-[[package]]
-name = "log"
-version = "0.1.10"
+version = "0.2.49"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-dependencies = [
- "regex 0.1.80 (registry+https://github.com/rust-lang/crates.io-index)",
-]

 [[package]]
 name = "log"
@@ -71,7 +63,7 @@ name = "memchr"
 version = "0.1.11"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
- "libc 0.2.48 (registry+https://github.com/rust-lang/crates.io-index)",
+ "libc 0.2.49 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
```

Running `cargo generate-lockfile` produce the same diff as `cargo update`

Reply:
> BTW if I clone that repo and run `cargo install` then it errors, e.g.
> 
> ```
> error[E0603]: module `stdio` is private
>    --> /Users/rhys/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.1.10/src/lib.rs:226:36
>     |
> 226 |     handle: LineBufferedWriter<io::stdio::StdWriter>,
>     |
> ```
> 
> Am I doing something wrong?

That repo has very old versions of crates as deps (`log 0.1.0` was released on December 14, 2014), so it most likely fails to build because recent versions of rust can't build them anymore.